src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,510 @@
+        /// <param name="properties">The properties.</param>
+        public MongoAWSAuthenticator(UsernamePasswordCredential credential, IEnumerable<KeyValuePair<string, string>> properties)
+            : this(credential, properties, new DefaultRandomByteGenerator(), SystemClock.Instance)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MongoAWSAuthenticator"/> class.
+        /// </summary>
+        /// <param name="username">The username.</param>
+        /// <param name="properties">The properties.</param>
+        public MongoAWSAuthenticator(string username, IEnumerable<KeyValuePair<string, string>> properties)
+            : this(username, properties, new DefaultRandomByteGenerator(), SystemClock.Instance)
+        {
+        }
+
+        internal MongoAWSAuthenticator(
+            UsernamePasswordCredential credential,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock dateTimeProvider)
+            : base(CreateMechanism(credential, properties, randomByteGenerator, dateTimeProvider))
+        {
+        }
+
+        internal MongoAWSAuthenticator(
+            string username,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock dateTimeProvider)
+            : base(CreateMechanism(username, null, properties, randomByteGenerator, dateTimeProvider))
+        {
+        }
+
+        /// <inheritdoc/>
+        public override string DatabaseName
+        {
+            get { return "$external"; }
+        }
+
+        // private static methods
+        private static MongoAWSMechanism CreateMechanism(
+            UsernamePasswordCredential credential,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock clock)
+        {
+            if (credential.Source != "$external")
+            {
+                throw new ArgumentException("MONGO AWS authentication may only use the $external source.", nameof(credential));
+            }
+
+            return CreateMechanism(credential.Username, credential.Password, properties, randomByteGenerator, clock);
+        }
+
+        private static MongoAWSMechanism CreateMechanism(
+            string username,
+            SecureString securePassword,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock clock)
+        {
+            var awsCredentialsCreators = new Func<AwsCredentials>[]
+            {
+                () => CreateAwsCredentialsFromMongoCredentials(username, securePassword, properties),
+                () => CreateAwsCredentialsFromEnvironmentVariables(),
+                () => CreateAwsCredentialsFromEcsResponse(),
+                () => CreateAwsCredentialsFromEc2Response()
```

Comment:
by @rstam 
This trick of having an array of creators seems overly complicated, Why can't this whole method just be:

    private static MongoAWSMechanism CreateMechanism(
        string username,
        SecureString securePassword,
        IEnumerable<KeyValuePair<string, string>> properties,
        IRandomByteGenerator randomByteGenerator,
        IClock clock)
    {
        var awsCredentials =
            CreateAwsCredentialsFromMongoCredentials(username, securePassword, properties) ??
            CreateAwsCredentialsFromEnvironmentVariables() ??
            CreateAwsCredentialsFromEcsResponse() ??
            CreateAwsCredentialsFromEc2Response();
        if (awsCredentials == null)
        {
            throw new ArgumentException("A MONGODB-AWS authenticator must have an access key ID.");
        }

        ValidateAwsCredentials(awsCredentials);

        var credentials = new UsernamePasswordCredential("$external", awsCredentials.Username, awsCredentials.Password);

        return new MongoAWSMechanism(credentials, awsCredentials.SessionToken, randomByteGenerator, clock);
    }

This might require some slight refactoring of the Create methods to only return null or a fully usable credential.

Reply:
I can implement it this way, but in this case I must revert my changes back after satisfying Dmitry's comments: https://github.com/MikalaiMazurenka/mongo-csharp-driver/pull/22/commits/2614e7537eed6af0fee5c5617aa9c62a41735fcc#r393308069

Because from spec ([link](https://github.com/mongodb/specifications/blob/master/source/auth/auth.rst#obtaining-credentials)):
> If a username is provided without a password (or vice-versa) or if only a session token is provided Drivers MUST raise an error

I just clarified that in AWS chat and it means what I expect it to mean:
> We must validate credentials after each retrieval.

Which means that we must validate credentials which we acquire. Thus, `ValidateAwsCredentials` must be moved back into `Create...` methods, which would contradict Dmitry's requirements.

Reply:
I think I suggested the current way with array to have debugging a bit simpler.